companion-api/api/user.go

```diff
@@ -1,54 +1,208 @@
 package api
 
 import (
-	"crypto/sha256"
-	"encoding/hex"
 	"net/http"
 
 	restful "github.com/emicklei/go-restful"
 	"github.com/mcluseau/autorizo/api"
+	"github.com/mcluseau/autorizo/auth"
+	"github.com/mcluseau/autorizo/companion-api/backend"
 )
 
 // Register provide a restful.WebService from this API
 func (cApi *CompanionAPI) registerUsers(ws *restful.WebService) {
 	ws.
-		Route(ws.POST("/user").
-			To(cApi.createOrUpdateUser).
-			Doc("Create or update user.").
+		Route(ws.POST("/user/{user-id}").
+			To(cApi.createUser).
+			Doc("Create a new user.").
 			Consumes("application/json").
-			Produces("application/json").
-			Reads(UserReq{}))
+			Param(ws.PathParameter("user-id", "identifier of the user").DataType("string")).
+			Reads(backend.UserData{}))
 
+	ws.
+		Route(ws.PUT("/user/{user-id}").
+			To(cApi.updateUser).
+			Doc("Update password of an existing user.").
+			Consumes("application/json").
+			Param(ws.PathParameter("user-id", "identifier of the user").DataType("string")).
+			Reads(backend.UserData{}))
+
+	ws.
+		Route(ws.PUT("/user/{user-id}/password").
+			To(cApi.updateUserPassword).
+			Doc("Update password of an existing user.").
+			Consumes("application/json").
+			Param(ws.PathParameter("user-id", "identifier of the user").DataType("string")).
+			Reads(backend.PasswordHashData{}))
+
+	ws.
+		Route(ws.PUT("/user/{user-id}/claims").
+			To(cApi.updateUserClaims).
+			Doc("Update claims of an existing user.").
+			Consumes("application/json").
+			Param(ws.PathParameter("user-id", "identifier of the user").DataType("string")).
+			Reads(auth.ExtraClaims{}))
+
+	ws.
+		Route(ws.DELETE("/user/{user-id}").
+			To(cApi.deleteUser).
+			Doc("Delete an existing user.").
+			Consumes("application/json").
+			Param(ws.PathParameter("user-id", "identifier of the user").DataType("string")))
+
+}
+
+func (cApi *CompanionAPI) createUser(request *restful.Request, response *restful.Response) {
+	defer func() {
+		if err := recover(); err != nil {
+			// unhandled error
+			api.WriteError(err.(error), response)
+		}
+	}()
+
+	id := request.PathParameter("user-id")
+	if len(id) == 0 {
+		response.WriteErrorString(http.StatusUnauthorized, "No user given.")
```

Comment:
it's not `Unauthorized`, it's `UnprocessableEntity` (https://tools.ietf.org/html/rfc4918#section-11.2)